### PR TITLE
doc: add new tutorial "Out of focus deblur filter"

### PR DESCRIPTION
**Pull request by VladKarpushin**
### This pullrequest adds new tutorial how to use out of focus deblur filter

In this tutorial you will learn:
•	what is a degradation image model
•	what is PSF of out-of-focus image
•	how to restore a blurred image
•	what is Wiener filter

Thanks to my great Linux/GitHub guru for making this contribution possible (Alexander Sashnov github.com/asashnov @asashnov)

**WIP**
```
force_builders=linux,Docs
test_modules=core
```

### CODE REVIEW

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
+﻿/**
+* @file out_of_focus_deblur_filter.cpp
+* @brief You will learn how to recover an out-of-focus image by Wiener filter
+* @author Karpushin Vladislav, karpushin@ngs.ru, https://github.com/VladKarpushin
+*/
+#include "opencv2/opencv.hpp"
```
Comment (alalek) [inline samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: `opencv2/opencv.hpp` is old compatibility header (deprecated for new code including OpenCV code itself). It is not actual anymore due huge amount of OpenCV modules.

Use headers from used OpenCV modules directly:
- opencv2/core.hpp
- opencv2/imgcodecs.hpp
- opencv2/highgui.hpp
- opencv2/imgproc.hpp

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
+﻿/**
+* @file out_of_focus_deblur_filter.cpp
+* @brief You will learn how to recover an out-of-focus image by Wiener filter
+* @author Karpushin Vladislav, karpushin@ngs.ru, https://github.com/VladKarpushin
```
Comment (alalek) [inline samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: > @file out_of_focus_deblur_filter.cpp

Remove this unnecessary useless line. Keep samples as short as possible.

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    int R = 53;
+    int snr = 5200;
+
+    string strInFileName = "original.JPG";
+
+    if (argc == 4)
+    {
+        strInFileName = argv[1];
+        R = atoi(argv[2]);
+        snr = atoi(argv[3]);
```
Comment (alalek) [inline samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: OpenCV samples should use `cv::CommandLineParser`.

File: doc/tutorials/core/table_of_content_core.markdown [hunk 4]
```diff
@@ -125,3 +125,13 @@ understanding how to manipulate the images on a pixel level.
...
+
+-   @subpage tutorial_out_of_focus_deblur_filter
+
+    *Languages:* C++
+
+    *Compatibility:* \> OpenCV 2.0
+
+    *Author:* Karpushin Vladislav
+
+    You will learn how to recover an out-of-focus image by Wiener filter.
```
Comment (alalek) [inline doc/tutorials/core/table_of_content_core.markdown]: Move tutorial from "core" to "imgproc". Image processing / filtering should go there.

File: doc/opencv.bib [hunk 5]
```diff
@@ -1016,3 +1016,17 @@ @INPROCEEDINGS{Ke17
...
+  author={Gonzalez, Rafael C and others},
+  year={1987},
+  publisher={Addison Wesley Publishing Company}
+}
+
+@ARTICLE{gruzman,
+  title={Цифровая обработка изображений в информационных системах},
+  author={Грузман, ИС and Киричук, ВС and Косых, ВП and Перетягин, ГИ and Спектор, АА},
+  year={2000},
+  publisher={Изд-во НГТУ Новосибирск}
```
Comment (alalek) [inline doc/opencv.bib]: Need to check that this is showed correctly.

**Comment by alalek** [comment]
As a documentation improvement this patch should [go](https://github.com/opencv/opencv/wiki/How_to_contribute#faq) into 3.4 branch. We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp (hunk 1 above)
Comment (VladKarpushin) [reply samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: I got it

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp (hunk 2 above)
Comment (VladKarpushin) [reply samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: I did not find "@file" Do you mean below line?
@include cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp

Comment (alalek) [reply samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: I mean "line 2" of the current file.

Comment (VladKarpushin) [reply samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: I got it

File: samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp (hunk 3 above)
Comment (VladKarpushin) [reply samples/cpp/tutorial_code/core/out_of_focus_deblur_filter/out_of_focus_deblur_filter.cpp]: I see

File: doc/tutorials/core/table_of_content_core.markdown (hunk 4 above)
Comment (VladKarpushin) [reply doc/tutorials/core/table_of_content_core.markdown]: ok, i am dealing with it

File: doc/opencv.bib (hunk 5 above)
Comment (VladKarpushin) [reply doc/opencv.bib]: checked, it looks in this way. Is it correctly?
bib

**Comment by alalek** [comment]
Images are very big. Could you please change images resolution or put some small sub-image (ROI) only?

**Comment by dkurt** [comment]
@VladKarpushin, May I ask you to check a sample? With a8e9a3a88df83ae118b54e975056c687ac3d30d2 commit we got the following result for image from the tutorial:
result

**Comment by VladKarpushin** [comment]
Hi. I can explain it. I tried to make the contribution size as small as possible, so all contributed images are strongly compressed. In order to deblur the image, you should process the original full-size image.

Below you can see he original full-size image.
dsc_0148_edgetape

Below you can see deblured image
dsc_0148_edgetape_r 53_snr 5200

**Comment by dkurt** [comment]
> The out-of-focus image on this page is a real world image. The out-of-focus was achieved manually by camera optics.

@VladKarpushin, Thank you! Am I right that we cannot use it as a general deblurring algorithm but it's designed specifically for out-of-focus camera frames?

**Comment by VladKarpushin** [comment]
Yes, you are right, this method is for out-of-focus camera frames only. Out-of-focus distortion corresponds to circle shape PSF (point spread function). For other types of distortion, you should use other PSF. For example, line shape PSF corresponds to motion blur distortion.

Resizing and compression remove high frequencies of a signal, this is why you can not recover the signal.

**Comment by Bleach665** [comment]
@VladKarpushin . For same reason can you give original image for motion deblur demo?
https://docs.opencv.org/3.4.5/motion_original.jpg

**Comment by VladKarpushin** [comment]
@Bleach665 
Hi. Below you can see the original full-size color image.
P1030513
P1030513_short_black
P1030513_short_white

**Comment by Mrmachiner** [comment]
> @Bleach665
> Hi. Below you can see the original full-size color image.
> P1030513
> P1030513_short_black
> P1030513_short_white

Image blackcar and whitecar size(874x616) not good sir :(

**Pull request merged**